### DISCUSSION TITLE
Broadcast proof of concept

### DISCUSSION THREAD

**Discussion by mhite** [MarByteBeep/StageLinq#18] [Ideas]
Has anyone done any preliminary research on this service? What might it be?

Could it be the "broadcast" feature of the Prime series mixer, but somehow exposed over TCP?

**Comment by honusz** [comment]
> Has anyone done any preliminary research on this service? What might it be?

My hypothesis: it's 10ch audio stream, stereo pairs for each of the four deck/layers, and a stereo pair for the master mixed output.

> Could it be the "broadcast" feature of the Prime series mixer, but somehow exposed over TCP?

In a manner of speaking. According to the press release:

> "Broadcast mode allows the main output mix to stream over USB audio channels 1/2 rather than the default 9/10."

I know QT (the c++ framework InMusic uses) has the ability to send/receive [multichannel audio data](https://doc.qt.io/qt-6/qaudioformat.html), so I suspect that's what we'll need to decode the streams.

Just need to find some software in the wild that uses this service, I haven't come across any yet.

**Reply by mhite** [reply]
@honusz - Wouldn't that be amazing!

Have you used "Timecode"? https://timecodesync.com/

It claims to do 10-channel audio routing.

**Reply by honusz** [reply]
Yeah, that's part of what leads me to believe that may be it.

That might also be wishful thinking though (that it's something cool like that).
One weird thing is that the players (sc5k/6k) offer the Broadcast service, not the mixer (at least, not that I can tell). I would have thought it would come from the mixer (certainly a mixed master output would).

Kind of fun to speculate about. Makes me want to go dumpster diving, Kevin Mitnick style, at InMusic for some spec docs, lol

**Reply by honusz** [reply]
> Have you used "Timecode"? https://timecodesync.com/
> It claims to do 10-channel audio routing.

@mhite sadly it turns out the audio routing is for the actual time code output, which I just learned the hard way by nearly blowing my speakers/eardrums by routing it out of system audio 🙉

**Reply by mhite** [reply]
> Kind of fun to speculate about. Makes me want to go dumpster diving, Kevin Mitnick style, at InMusic for some spec docs, lol

@honusz - I agree! 💯

**Reply by mhite** [reply]
> > Have you used "Timecode"? https://timecodesync.com/
> > It claims to do 10-channel audio routing.
> 
> @mhite sadly it turns out the audio routing is for the actual time code output, which I just learned the hard way by nearly blowing my speakers/eardrums by routing it out of system audio 🙉

@honusz - Ouch!

**Comment by honusz** [comment]
For a lark, I tried connecting to the Broadcast service in the manner of any other service. The device accepts the connection, but doesn't send any data (even if a track is playing). 
There must be some sort of client-side message required to initiate... well, whatever it does.

**Reply by mhite** [reply]
> There must be some sort of client-side message required to initiate... well, whatever it does.

@honusz - I wish I knew! It would be great to track down what Stagelinq application actually uses this.

**Comment by honusz** [comment]
So, managed to find some Broadcast packets by chance while investigating OfflineAnalyzer.
@mhite prepare to be underwhelmed.....

It seems like this service just broadcasts changes in the Database to a device that's presumably downloaded the DB.

Some sample messages:

`{"engine.track.changed":{"databaseUuid":"327ea661-af29-4b6e-ba38-b1c6bd2a7879","trackId":"340"}}`
`{"engine.historylist.added":{"databaseUuid":"20810c18-576a-48b5-b1af-23c0d6b6457e","listId":"27","sessionId":"684769930"}}`

Very useful things for the players to share, I'm sure. But not a lot a lot of sex in this service.

To connect, you just have to connectToService as you do any other service, theres no reply, it just sends data when theres an update (I guess I wasn't patient enough).
Off to play around with it a bit more, I'll update as I learn more.

**Comment by mhite** [comment]
@honusz - Still, this is pretty delicious! 

I hope to play with this soon, although I wonder what / if anything will show up on a Prime 4.

**Reply by mhite** [reply]
@honusz - Ok, I played around a bit with this tonight and started implementing it in the @icedream library. FWIW, with my Prime4, I saw the same two types of messages you witnessed. (`engine.track.changed` and `engine.historylist.added`)

**Reply by honusz** [reply]
Yeah, that's pretty much it.

I got some performance data changed messages as well, I believe that was while it was encoding.
I believe these messages triggered the other player to download a a db entry for the track, but I still need to sift through those captures (they're huge, lol)

**Reply by mhite** [reply]
I've got proof of concept working:

```
./broadcast
2022/10/04 22:00:07 Listening for devices for 5s
2022/10/04 22:00:07 192.168.68.198 "prime4" "OfflineAnalyzer" "1.0.0"
2022/10/04 22:00:07 	attempting to connect to this device...
2022/10/04 22:00:07 	requesting device data services…
2022/10/04 22:00:07 	offers FileTransfer at port 46285
2022/10/04 22:00:07 	end of list of device data services
2022/10/04 22:00:07 192.168.68.198 "prime4" "JC11" "2.3.2"
2022/10/04 22:00:07 	attempting to connect to this device...
2022/10/04 22:00:07 	requesting device data services…
2022/10/04 22:00:08 	offers StateMap at port 44403
2022/10/04 22:00:08 	offers Broadcast at port 41291
2022/10/04 22:00:08 		connecting to Broadcast...
2022/10/04 22:00:08 		PRESS CTRL-C TO ABORT!
2022/10/04 22:00:46 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:135]]}
2022/10/04 22:00:46 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:157]]}
2022/10/04 22:00:46 			{Value:map[engine.historylist.trackplayed:map[databaseUuid:ae49c6d4-a769-4129-9f1d-8a2d36ee7cfa listId:161 sessionId:957276061]]}
2022/10/04 22:00:46 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:140]]}
2022/10/04 22:00:46 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:180]]}
2022/10/04 22:00:46 			{Value:map[engine.historylist.trackplayed:map[databaseUuid:ae49c6d4-a769-4129-9f1d-8a2d36ee7cfa listId:161 sessionId:957276061]]}
2022/10/04 22:05:54 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:29]]}
2022/10/04 22:05:54 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:90]]}
2022/10/04 22:05:54 			{Value:map[engine.historylist.trackplayed:map[databaseUuid:ae49c6d4-a769-4129-9f1d-8a2d36ee7cfa listId:161 sessionId:957276061]]}
2022/10/04 22:05:54 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:37]]}
2022/10/04 22:05:54 			{Value:map[engine.track.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:89]]}
2022/10/04 22:05:54 			{Value:map[engine.historylist.trackplayed:map[databaseUuid:ae49c6d4-a769-4129-9f1d-8a2d36ee7cfa listId:161 sessionId:957276061]]}
2022/10/04 22:07:48 			{Value:map[engine.device.ejected:map[deviceName:Music (USB 1)]]}
```

I haven't seen a performance data changed message -- what's that look like? How did you trigger it? (Not sure what you mean "while it was encoding")

Only thing I have to add is that I triggered an eject event -- `{Value:map[engine.device.ejected:map[deviceName:Music (USB 1)]]}`

**Reply by mhite** [reply]
Here's another:

```
2022/10/04 22:13:22 			{Value:map[engine.preparelist.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1]]}
2022/10/04 22:13:24 			{Value:map[engine.preparelist.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1]]}
```

If you hit shift along with the track selection knob on a Prime 4 while browsing a playlist, this places a track in the prepare list. The above events show up when you do that.

**Reply by mhite** [reply]
Just doing some dumpster diving in binaries, it seems like these events also trigger, but I have no idea the structure of them:

```
engine.craylist.structureChanged
engine.craylist.contentChanged
engine.historylist.added
```

Have you spotted any of these in captures?

**Reply by mhite** [reply]
Ok, actually, I triggered a performance data change message when I shutdown my Prime4.

```
{Value:map[engine.performanceData.changed:map[databaseUuid:d3ba795c-2f84-4164-a1b6-27b9348847b1 trackId:103]]}
```

**Reply by honusz** [reply]
> I haven't seen a performance data changed message -- what's that look like? How did you trigger it? (Not sure what you mean "while it was encoding")
Whoops, I should have said "while it was _analyzing_" as I was capturing with a Rekordbox usb connected to the device (this was when I was testing the OfflineAnalyzers).

`{"engine.performanceData.changed":{"databaseUuid":"                                   a9d94ebc-2f81-4447-9306-fbbd1588540a","trackId":"%2"}}`

Interesting that the trackId is %2, not sure what that means. A lot of these messages also have a strange amount of excess 0x20 ascii characters in between `databaseUuid":"` and the start of the UUID. 
Even in the post preview these disappear, so it's hard to see what I mean. Heres the hex dump of the message:
```0000   00 05 95 02 33 f9 00 05 95 02 34 0c 08 00 45 00   ....3.....4...E.
0010   00 c4 86 ff 40 00 40 06 2d 3d c0 a8 02 54 c0 a8   ....@.@.-=...T..
0020   02 53 b7 05 93 f3 15 39 67 79 20 0c c8 c7 80 18   .S.....9gy .....
0030   01 f5 5d 17 00 00 01 01 08 0a 82 fa fa b5 ed f3   ..].............
0040   e8 7f 00 00 00 8c 7b 22 65 6e 67 69 6e 65 2e 70   ......{"engine.p
0050   65 72 66 6f 72 6d 61 6e 63 65 44 61 74 61 2e 63   erformanceData.c
0060   68 61 6e 67 65 64 22 3a 7b 22 64 61 74 61 62 61   hanged":{"databa
0070   73 65 55 75 69 64 22 3a 22 20 20 20 20 20 20 20   seUuid":"       
0080   20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20                   
0090   20 20 20 20 20 20 20 20 20 20 20 20 61 39 64 39               a9d9
00a0   34 65 62 63 2d 32 66 38 31 2d 34 34 34 37 2d 39   4ebc-2f81-4447-9
00b0   33 30 36 2d 66 62 62 64 31 35 38 38 35 34 30 61   306-fbbd1588540a
00c0   22 2c 22 74 72 61 63 6b 49 64 22 3a 22 25 32 22   ","trackId":"%2"
00d0   7d 7d                                             }}
```

Have you seen that?

**Reply by mhite** [reply]
I have not seen anything like that! It sure looks like an EngineOS bug, though. `%2` is just a field position element you'd normally see in a `printf` statement. Also that `databaseUuid` field having a bunch of extra spaces (20 hex, 32 decimal) seems like another bug... almost like EngineOS should have done a left trim on that UUID before crafting the payload. I'll keep my eye out for this weirdness!
